sql/logictest: TestLogic failed

sql/logictest.TestLogic [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3422600&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3422600&tab=artifacts#/) on master @ [5b40b1920e20b94ded55cea9a16a5a7a6adc7157](https://github.com/cockroachdb/cockroach/commits/5b40b1920e20b94ded55cea9a16a5a7a6adc7157):


```
=== RUN   TestLogic
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic113198071
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestLogic
    logic.go:3424: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestLogic113198071
--- FAIL: TestLogic (2394.32s)
=== RUN   TestLogic/fakedist-disk/select
=== PAUSE TestLogic/fakedist-disk/select
=== CONT  TestLogic/fakedist-disk/select
--- done: testdata/logic_test/collatedstring_index1 with config fakedist-disk: 19 tests, 0 failures
=== CONT  TestLogic/fakedist-disk/select
    logic.go:2562: 
        
        testdata/logic_test/select:769: 
        expected success, but found
        (P0003) query ``SELECT * FROM t_disallow_scans`` contains a full table/index scan which is explicitly disallowed
        conn_executor_exec.go:1238: in makeExecPlan()
--- done: testdata/logic_test/select with config fakedist-disk: 165 tests, 1 failures
E210910 09:15:08.913623 8455200 sql/logictest/logic.go:3657  [-] 853  pq: query ``SELECT * FROM t_disallow_scans`` contains a full table/index scan which is explicitly disallowed
    logic.go:3192: 
        testdata/logic_test/select:769: error while processing
    logic.go:3192: pq: query ``SELECT * FROM t_disallow_scans`` contains a full table/index scan which is explicitly disallowed
        --- FAIL: TestLogic/fakedist-disk/select (1.27s)
=== RUN   TestLogic/fakedist-disk
    --- FAIL: TestLogic/fakedist-disk (0.01s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestLogic PKG=./pkg/sql/logictest TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=2
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #63262 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-20.2]
- #61657 sql/logictest: TestLogic failed [C-test-failure O-robot branch-release-21.1]
- #58269 sql/logictest: TestLogic failed [C-test-failure O-robot T-sql-schema branch-release-20.1]
</p>
</details>
/cc @cockroachdb/sql-queries jordanlewis
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/logictest/testdata/logic_test/select
+++ b/pkg/sql/logictest/testdata/logic_test/select
@@ -725,8 +725,10 @@ true NULL
 
 # Tests for `disallow_full_table_scans`
 statement ok
+SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false
+
+statement ok
 CREATE TABLE t_disallow_scans(a INT, b INT, INDEX b_idx(b));
-INSERT INTO t_disallow_scans VALUES (1, 1), (2, 1), (3, 1)
 
 statement ok
 SELECT * FROM t_disallow_scans
@@ -741,11 +743,35 @@ SELECT * FROM t_disallow_scans
 statement error pq: query `SELECT \* FROM t_disallow_scans@b_idx` contains a full table/index scan which is explicitly disallowed
 SELECT * FROM t_disallow_scans@b_idx
 
-# Disable 'large_full_scan_rows' heuristic and analyze the table in order to
+# Disable 'large_full_scan_rows' heuristic and inject the stats in order to
 # test 'disallow_full_table_scans' in isolation.
 statement ok
 SET large_full_scan_rows = 0;
-ANALYZE t_disallow_scans
+
+statement ok
+ALTER TABLE t_disallow_scans INJECT STATISTICS '[
+  {
+    "columns": ["rowid"],
+    "created_at": "2021-09-15 19:38:46.017315",
+    "distinct_count": 3,
+    "null_count": 0,
+    "row_count": 3
+  },
+  {
+    "columns": ["b"],
+    "created_at": "2021-09-15 19:38:46.017315",
+    "distinct_count": 1,
+    "null_count": 0,
+    "row_count": 3
+  },
+  {
+    "columns": ["a"],
+    "created_at": "2021-09-15 19:38:46.017315",
+    "distinct_count": 3,
+    "null_count": 0,
+    "row_count": 3
+  }
+]'
 
 statement error pq: query `SELECT \* FROM t_disallow_scans` contains a full table/index scan which is explicitly disallowed
 SELECT * FROM t_disallow_scans
@@ -762,7 +788,7 @@ SELECT * FROM crdb_internal.jobs
 
 # Tests for 'large_full_scan_rows'.
 statement ok
-SET large_full_scan_rows = 4
+SET large_full_scan_rows = 1000
 
 # These queries succeed because the full table/index scans aren't considered
 # "large".
@@ -784,7 +810,8 @@ SELECT * FROM t_disallow_scans@b_idx
 # Cleanup
 statement ok
 SET disallow_full_table_scans = false;
-RESET large_full_scan_rows
+RESET large_full_scan_rows;
+RESET CLUSTER SETTING sql.stats.automatic_collection.enabled;
 
 # Regression test for #58104.
 statement ok

